service/src/main/java/org/databiosphere/workspacedataservice/dataimport/PfbQuartzJob.java

```diff
@@ -27,7 +48,49 @@ protected JobDao getJobDao() {
 
   @Override
   protected void executeInternal(UUID jobId, JobExecutionContext context) {
-    // TODO: AJ-1227? implement PFB import.
+    JobDataMap jobDataMap = context.getMergedJobDataMap();
+    URL url = getJobDataUrl(jobDataMap, ARG_URL);
+    try (DataFileStream<GenericRecord> dataStream =
+        PfbReader.getGenericRecordsStream(url.toString())) {
+      // translate the Avro DataFileStream into a Java stream
+      Stream<GenericRecord> recordStream =
+          StreamSupport.stream(
+              Spliterators.spliteratorUnknownSize(dataStream.iterator(), Spliterator.ORDERED),
+              false);
+
+      // process the stream into a list of unique snapshotIds
+      List<String> snapshotIds =
+          recordStream
+              .map(rec -> rec.get("object")) // Records in a pfb are stored under the key "object"
+              .filter(GenericRecord.class::isInstance) // which we expect to be a GenericRecord
+              .filter(
+                  obj ->
+                      ((GenericRecord) obj)
+                          .hasField(SNAPSHOT_ID_IDENTIFIER)) // avoid exception if field nonexistent
+              .map(
+                  obj ->
+                      ((GenericRecord) obj)
+                          .get(SNAPSHOT_ID_IDENTIFIER)
+                          .toString()) // within the GenericRecord, find the
+              // source_datarepo_snapshot_id
+              .filter(Objects::nonNull) // expect source_datarepo_snapshot_id to be non-null
+              .distinct() // find only the unique snapshotids
+              .toList();
+
+      // TODO AJ-1371 pass snapshotIds to WSM
+      for (String id : snapshotIds) {
+        try {
+          wsmDao.createDataRepoSnapshotReference(new SnapshotModel().id(UUID.fromString(id)));
+        } catch (Exception e) {
+          throw new PfbParsingException("Error processing PFB: Invalid snapshot UUID");
+        }
+      }
+
+    } catch (IOException e) {
+      throw new PfbParsingException("Error processing PFB");
```

Comment:
When catching/re-raising I usually try to serve up the original exception too, otherwise the caller has to go look into the sourcecode to figure out what happened.

Can you make `PfbParsingException` allow a wrapped exception and then use it like so?

```suggestion
      throw new PfbParsingException("Error processing PFB", e);
```